### PR TITLE
feat: support UnixFSv1.5 metadata

### PR DESCRIPTION
**Pull request by achingbrain**
Depends on:

- [x] https://github.com/ipfs/js-ipfs-utils/pull/19
- [x] https://github.com/ipfs/interface-js-ipfs-core/pull/572

### CODE REVIEW

File: src/add/form-data.js [hunk 1]
```diff
@@ -12,6 +12,14 @@ exports.toFormData = async input => {
   let i = 0
 
   for await (const file of files) {
+    if (file.mtime) {
+      formData.append(`file-${i}-mtime`, file.mtime)
```
Comment (alanshaw) [inline src/add/form-data.js]: What if this is `dir-${i}`?

File: src/add/form-data.browser.js [hunk 2]
```diff
@@ -9,6 +9,14 @@ exports.toFormData = async input => {
   let i = 0
 
   for await (const file of files) {
+    if (file.mtime) {
+      formData.append('mtime', file.mtime)
```
Comment (alanshaw) [inline src/add/form-data.browser.js]: Need to identify which file/dir this metadata pertains to.

File: src/add/form-data.js (hunk 1 above)
Comment (achingbrain) [reply src/add/form-data.js]: The field names aren't used when parsing multipart requests - if it's a dir we detect that from the content type.

File: src/add/form-data.browser.js (hunk 2 above)
Comment (achingbrain) [reply src/add/form-data.browser.js]: This has been refactored out.

**Comment by achingbrain** [comment]
Adding interface tests are blocked by https://github.com/ipfs/js-ipfs/pull/2625

File: src/add/form-data.browser.js [hunk 3]
```diff
@@ -18,9 +28,13 @@ exports.toFormData = async input => {
...
 
-      formData.append(`file-${i}`, new Blob(bufs, { type: 'application/octet-stream' }), encodeURIComponent(file.path))
+      formData.append(`file-${i}`, new Blob(bufs, { type: 'application/octet-stream' }), encodeURIComponent(file.path), {
+        header: headers
+      })
     } else {
-      formData.append(`dir-${i}`, new Blob([], { type: 'application/x-directory' }), encodeURIComponent(file.path))
+      formData.append(`dir-${i}`, new Blob([], { type: 'application/x-directory' }), encodeURIComponent(file.path), {
+        header: headers
+      })
```
Comment (alanshaw) [inline src/add/form-data.browser.js]: I'm not sure this is working...

```js
const fd = new FormData
fd.append('test', new Blob(['data'], { type: 'application/octet-stream' }), 'path', { header: { mtime: Date.now(), mode: 'rwx' } })

const opts =  { method: 'post', body: fd }
const r0 = new Request('/test', opts)

await r0.text()
"-----------------------------21291924801818001399752183477

Content-Disposition: form-data; name=\"test\"; filename=\"path\"

Content-Type: application/octet-stream



data

-----------------------------21291924801818001399752183477--

"
```

File: src/add/form-data.browser.js [hunk 4]
```diff
@@ -9,6 +9,16 @@ exports.toFormData = async input => {
...
 
   for await (const file of files) {
+    const headers = {}
+
+    if (file.mtime) {
+      headers.mtime = file.mtime
+    }
+
+    if (file.mode) {
+      headers.mode = file.mode
```
Comment (alanshaw) [inline src/add/form-data.browser.js]: Missing?

```suggestion
      headers.mode = file.mode.toString(8).padStart(4, '0')
```

**Comment by achingbrain** [comment]
I'm going to merge this to stop it getting stale - interface tests are in & skipped for go-IPFS, we can address any issues in further (smaller) PRs.

**Pull request merged**